File: lee/src/main/java/org/example/lee/题目/链表/b_环形链表Ⅱ.go
package 链表

// 数学题
// 假设头部到环头部的距离为a,环头部到相遇点距离为b,相遇点到环头部距离为c
// 假设 一快一慢指针 快是慢的2倍速度
// 直接假设 慢指针入环后一圈没转就被追上
// 有 2(a+b) = a+b+c+b
// 解得 a = c
// 代码模拟即可
func detectCycle(head *ListNode) *ListNode {
	a, b := head, head
	for a != nil && b != nil && b.Next != nil && a != b {
		a = a.Next
		b = b.Next.Next
	}
	if a == nil || b == nil || b.Next == nil {
		return nil
	}
	for head != a {
		a = a.Next
		head = head.Next
	}
	return head
}

File: lee/src/main/java/org/example/lee/题目/链表/b_环形链表Ⅱ.java
package org.example.lee.题目.链表;

import org.example.lee.model.ListNode;

public class b_环形链表Ⅱ {

	/**
	 * detect cycle
	 * https://www.programmercarl.com/0142.%E7%8E%AF%E5%BD%A2%E9%93%BE%E8%A1%A8II.html#%E6%80%9D%E8%B7%AF
	 * 数学题 需要自己列一下相关式子 然后就能推导出结论
	 *
	 * @param head head
	 * @return {@link ListNode}
	 */
	public ListNode detectCycle(ListNode head) {
		if (head == null || head.next == null || head.next.next == null) {
			return null;
		}
		ListNode a = head.next, b = head.next.next;
		while (a != null && b != null && b.next != null && a != b) {
			b = b.next.next;
			a = a.next;
		}
		if (a == null || b == null || b.next == null) {
			return null;
		}
		while (head != a) {
			head = head.next;
			a = a.next;
		}
		return a;
	}
}
